### ISSUE TITLE
Transferring blobs / binary data

### ISSUE THREAD

**Issue by nanomonkey** [taoensso/sente#381]
Many thanks for Sente, I've been happily using it for some time.

Are there any examples of sending and receiving binary files, such as images or database blobs?  I'm unsure if there is a more appropriate route than base64 encoding/decoding.  Is it possible to open up a binary websocket for such messages?

Thanks again!

**Comment by ptaoussanis** [comment]
Hey there, apologies for the slow response!

Sente unfortunately doesn't currently provide anything specifically for binary messages out-the-box.
Haven't given it much thought, would be open to seeing ideas/PRs 👍 

Otherwise base64 would be one option in the meantime.

Cheers!

**Comment by awb99** [comment]
Binary blobs would be cool. Especially when sending larger payload data there is a significant overhead with the encoding. Now that say Apache aeriw can be sent as blobs via http get it would be cool to be able to send them via sente websocket as well.

**Comment by rosejn** [comment]
Hi, so I've implemented binary support, and unlike some of the previous discussions around what this would take, it only required minor modifications of about 10 lines of code, not including the packers.  I'm using this with msgpack in both Clojure and Clojurescript, and I've extended it to work with typed arrays on both ends, while also supporting all of the Clojure data structures and base value types in binary form.  Binary blobs, files, or anything else would be trivial to add.  For now though, I'd like to try to get these modifications accepted, and then I'm happy to also contribute the msgpack implementation if you'd like to include that in sente.  Those require 2 external libraries, although if I get the time I'd like to unify them and make a new msgpack.cljc library that cleanly implements it once and for all.  We needed this ASAP for a product release though, so for now here we are.

Here's the whole diff:
https://github.com/rosejn/sente/commit/3de8274f7451d402d342f327b174672eaf0724e3

In short, the current code wraps the string to transmit over the wire with an optional + or - and a callback UUID in case the user hopes for an RPC style reply.  That was the only code which relied on the msg being a string.  I changed that to instead wrap the msg in a clojure vector which does the same thing, except this way the packers are totally responsible for serialization.  If you have thoughts on another approach I'm happy to modify this.  Thanks for Sente,

**Comment by ptaoussanis** [comment]
@rosejn Thanks Jeff, that sounds promising! Would be very happy to see a PR. Just a heads-up that I'd only expect to review+merge next time I'm doing some batched work on Sente.

Cheers!

**Comment by lynaghk** [comment]
Hey y'all, I found https://github.com/rosejn/msgpack-cljc via this thread and gave it a shot with Sente, but the `extend-msgpack` referenced by @rosejn in [his Sente PR](https://github.com/taoensso/sente/pull/398#issuecomment-1142589850)
 for handling custom types doesn't actually work in ClojureScript (where a few tags are hardcoded in [this case statement](https://github.com/rosejn/msgpack-cljc/blob/e29d9d07b1e8b6947326f169aa37ba1f2521142c/src/msgpack/pack.cljs#L325-L338))

Issues are disabled on the msgpack-cljc repo, so I figured I'd raise it here for any other folks looking at sending binary data between Clojure/Clojurescript via Sente.

I'll have a go at forking msgpack-cljc so that the extension mechanism works on cljs. @ptaoussanis please let me know if there are any considerations or alternative binary formats you want me to investigate.

**Comment by ptaoussanis** [comment]
@lynaghk Hi Kevin! Thanks for pinging about this, and for the info 🙏

I'm actually about to switch focus to Sente for a bit, want to aim for a new release around the end of June. As part of that, am planning to investigate the possibility of adding support for WebSocket content compression. Could be I end up changing some of the relevant Sente APIs.

This could be a good time to also consider possible changes/additions/etc. re: binary data handling, but I don't have any thoughts yet.

> @ptaoussanis please let me know if there are any considerations or alternative binary formats you want me to investigate.

I'm not up to date on current possibilities, but I'd be happy to hear suggestions!

**Comment by lynaghk** [comment]
Compression sounds useful. Please make it optional, as some use cases might be faster without it (e.g., sending data to a browser on localhost).

Re: alternative formats, I did a bit of digging and think I'll keep using MessagePack. This [recent blog post](https://news.ycombinator.com/item?id=43978476) benchmarks a few formats. I think the main reason to use something besides MessagePack would be performance. I'd probably try flatbuffer first for its selective decoding and minimal size (out of band schema), but I doubt that's a common need for typical Sente and Clojure folks so won't pursue it for now.

**Comment by rosejn** [comment]
Hey guys,
  It's been a few years since I worked on this, but in that project they have been using this code for msgpack between clojure and clojurescript with custom types for years, so it definitely works in some shape or form.  Here are the two relevant repos, which might help:

https://github.com/rosejn/sente
https://github.com/rosejn/msgpack-cljc

Sorry I don't have time to dig in further right now, but hopefully this helps.

-Jeff

**Comment by ptaoussanis** [comment]
@lynaghk

> Please make it optional, as some use cases might be faster without it (e.g., sending data to a browser on localhost).

👍

@rosejn Thank you Jeff, that's helpful! Hope you're well, cheers :-)

**Comment by ptaoussanis** [comment]
Okay, quick update on this:

1. I think this _particular_ issue (related to supporting binary packers) can probably be closed.
2. @lynaghk It'd be great if you wanted to take care of either looking at an updated fork of [msgpack-cljc](https://github.com/rosejn/msgpack-cljc) or pursuing some other format of your choice. Please feel free to open a new issue for that when convenient 🙏

From my end I'll see about possibly:

- Adding support for binary un/packing for Ajax channel sockets
- Adding support for async un/packing, which should be handy for optional client-side compression
